### PR TITLE
Add reshape and flatten for dask.array

### LINKED ISSUE #570: dask.array flatten and reshape
The numpy method `flatten` can be used to reshape an n-dimensional array into a one-dimensional array.

```python
In [1]: import numpy as np

In [2]: x = np.array([[1, 2, 3], [4, 5, 6]])

In [3]: x
Out[3]: 
array([[1, 2, 3],
       [4, 5, 6]])

In [4]: x.flatten()
Out[4]: array([1, 2, 3, 4, 5, 6])
```

We could add such a method onto `dask.array.Array` if we are willing to give up caring about byte order entirely.  

This might be a good on-ramp issue.  It requires applying the `flatten` method onto each block, creating a new flattened dictionary, manipulating `chunks`, etc..  Furthermore we could use this as a learning opportunity to improve documentation on how to [extend dask array](http://dask.pydata.org/en/latest/array-extend.html).

cc @kwmsmith

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #570

Currently only does `order='C'`. A future PR might add the option for `order='K'`, which could be useful in some cases.

Example usage:
```
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: a = da.from_array(np.random.randint(10, size=(3, 4)), chunks=2)

In [4]: a.reshape(2, 6).compute()
Out[4]:
array([[7, 9, 6, 4, 2, 7],
       [7, 8, 3, 5, 4, 7]])

In [5]: a.reshape(-1).compute()
Out[5]: array([7, 9, 6, 4, 2, 7, 7, 8, 3, 5, 4, 7])

In [7]: a.ravel().compute()
Out[7]: array([7, 9, 6, 4, 2, 7, 7, 8, 3, 5, 4, 7])
```

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -2095,6 +2107,65 @@ def broadcast_to(x, shape):
...
+    known_sizes = [s for s in shape if s != -1]
+    if len(known_sizes) < len(shape):
+        if len(known_sizes) - len(shape) > 1:
+            raise ValueError('can only specify one unknown dimension')
+        missing_size = int(array.size / np.prod(known_sizes))
+        shape = tuple(missing_size if s == -1 else s for s in shape)
+
+    if np.prod(shape) != array.size:
+        raise ValueError('total size of new array must be unchanged')
+    return unravel(ravel(array), shape)
```
Comment (mrocklin) [inline dask/array/core.py]: Hrm, clever

**Comment by mrocklin** [comment]
Have you tried this on real-world problems?  I'm a bit concerned that this might result in very many tasks.

**Comment by shoyer** [comment]
The compelling use case for this PR is reshaping a stack of images along the last axes. For this use case, the shortcut logic in this PR is actually per efficient.

Notice how compact these graphs are:
```python
import dask.array as da
x = da.ones((50, 1000, 1000), chunks=(10, 1000, 1000))
x.ravel().visualize()
```
image
```python
x.reshape(50, -1).visualize()
```
image

On the other hand, if we try to reshape along the wrong axis, things get much messier:
```python
da.ones((5, 10, 10), chunks=5).reshape(5, -1).visualize()
```
image
Given how bad this looks, it might be better to simply disallow this sort of reshaping.

Note that the "stack of images" is probably the most compelling use case for dask.array right now -- I would put climate data use cases such as "one big file stores each month's data" under that same category. The efficient types of reshaping that are allowed here would be enough to allow for arbitrary raveling/unraveling of dimensions in xray in many cases (e.g., converting a 2D array with dimensions ('x', 'y') to a 1D array with a multi-index and vise-versa).

**Comment by mrocklin** [comment]
OK, sounds like we should merge this and wait to see if it gets misused.  If misuse ever occurs then maybe we look into raising warnings at the appropriate times. 

+1

**Comment by jeremyherbert** [comment]
Hi,

I'd just like to add that I think I have tripped on the problem you were worrying about.


setup:
```
window_size = 1e6
number_of_windows = 7e5

data = da.from_array(d, chunks=(1e6, ))
```


```
# this takes forever and crashes 
windowed = data[:window_size*number_of_windows].reshape((-1, window_size))
```
and the result:
```
Traceback (most recent call last):
  File "/Users/jeremy/Google Drive/phd/downsample_plot/downsample.py", line 81, in <module>
    real_window_size))
  File "/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py", line 890, in reshape
    return reshape(self, shape)
  File "/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py", line 2224, in reshape
    return unravel(ravel(array), shape)
  File "/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py", line 2209, in unravel
    for n in range(shape[0])])
  File "/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py", line 1494, in stack
    assert len(set(a.chunks for a in seq)) == 1  # same chunks
AssertionError
```

```
# this appears to work and is instant
windowed = data[:window_size*number_of_windows].reshape((number_of_windows, window_size))
```

**Comment by shoyer** [comment]
Those lines of code should do exactly the same thing, so there might be a bug lurking here. Does it work if you cast the sizes to integers with int?

**Comment by jeremyherbert** [comment]
Hi @shoyer,

I can't get my code to break anymore unfortunately, I'm guessing it was something I did wrong. But I did find a problem with float handling, and I have created a new issue for it.

**Comment by cowlicks** [comment]
Great work @shoyer

**Comment by mrocklin** [comment]
Possible fail case due to many-small-tasks: http://stackoverflow.com/questions/37912827/dask-array-reshape-very-slow

I don't think we should really do anything so resolve the problem (other than maybe raise an informative warning.)  Just pointing it out.

**Pull request merged**